contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java

```diff
@@ -0,0 +1,615 @@
+
+  private final RunningSpanStore runningSpanStore;
+  private final SampledSpanStore sampledSpanStore;
+
+  private TracezPageFormatter(
+      @Nullable RunningSpanStore runningSpanStore, @Nullable SampledSpanStore sampledSpanStore) {
+    this.runningSpanStore = runningSpanStore;
+    this.sampledSpanStore = sampledSpanStore;
+  }
+
+  /**
+   * Constructs a new {@code TracezPageFormatter}.
+   *
+   * @param runningSpanStore the instance of the {@code RunningSpanStore} to be used.
+   * @param sampledSpanStore the instance of the {@code SampledSpanStore} to be used.
+   * @return a new {@code TracezPageFormatter}.
+   */
+  public static TracezPageFormatter create(
+      @Nullable RunningSpanStore runningSpanStore, @Nullable SampledSpanStore sampledSpanStore) {
+    return new TracezPageFormatter(runningSpanStore, sampledSpanStore);
+  }
+
+  /**
+   * Emits the HTML generated page to the {@code outputStream}.
+   *
+   * @param queryMap the query components map.
+   * @param outputStream the output {@code OutputStream}.
+   */
+  public void emitHtml(Map<String, String> queryMap, OutputStream outputStream) {
+    PrintWriter out =
+        new PrintWriter(
+            new BufferedWriter(new OutputStreamWriter(outputStream, Charset.forName("US-ASCII"))));
+    out.write("<html><head><title>Tracez</title>");
+    out.write("<link rel=\"shortcut icon\" href=\"//www.opencensus.io/favicon.ico\"/>");
+    out.write("</head>\n");
+    if (runningSpanStore == null || sampledSpanStore == null) {
+      out.write("OpenCensus implementation not available.");
+      out.close();
+      return;
+    }
+    Formatter formatter = new Formatter(out, Locale.US);
+    emitSummaryTable(out, formatter);
+    String spanName = queryMap.get(HEADER_SPAN_NAME);
+    if (spanName != null) {
+      tracer
+          .getCurrentSpan()
+          .addAnnotation(
+              "Render spans.",
+              ImmutableMap.<String, AttributeValue>builder()
+                  .put("SpanName", AttributeValue.stringAttributeValue(spanName))
+                  .build());
+      String typeStr = queryMap.get(HEADER_SAMPLES_TYPE);
+      if (typeStr != null) {
+        List<SpanData> spans = null;
+        int type = Integer.parseInt(typeStr);
+        if (type < 0 || type > TYPE_ERROR) {
+          return;
+        }
+        if (type == TYPE_RUNNING) {
+          // Display running.
+          spans =
+              new ArrayList<>(
+                  runningSpanStore.getRunningSpans(RunningSpanStore.Filter.create(spanName, 0)));
+          // Sort active spans incremental.
+          Collections.sort(spans, new SpanDataComparator(true));
+        } else {
+          String subtypeStr = queryMap.get(HEADER_SAMPLES_SUB_TYPE);
+          if (subtypeStr != null) {
+            int subtype = Integer.parseInt(subtypeStr);
+            if (type == TYPE_ERROR) {
+              if (subtype < 0 || subtype >= CanonicalCode.values().length) {
+                return;
```

Comment:
Same. I would propose to put the "meat" of HTML generation into a separate method and throw in case of errors. The "parent" method will generate HTML header/tail and will also produce a nice error if/when exception is thrown.